dali/kernels/reduce/reduce_drop_dims.h

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_DROP_DIMS_H_
+#define DALI_KERNELS_REDUCE_REDUCE_DROP_DIMS_H_
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <stdexcept>
+#include "dali/core/util.h"
+#include "dali/core/host_dev.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+/**
+ * Calculates a flat index in a tensor with reduced dimensions based on the
+ * index in original tensor.
+ *
+ * This is used in calculation of standard deviation when subtracting mean.
+ *
+ * @details When reducing non-adjacent dimensions, the outer reduction will see
+ * all inner dimensions flattened - reduced and non-reduced alike. This flat index
+ * cannot be used to index the the tensor of means, which has more reduced dimensions.
+ * Example - calculating mean/stddev of all pixel brightness values in a mutli-channel video:
+ * input layout FHWC
+ * output layout 1HW1
+ * When calculating variance in F, we still see HWC as the (fused) inner dimension, but it is
+ * different than HW1 in the tensor of means. This class implements this kind of reindexing.
+ *
+ * The reindexing is done by either dividing and multiplying by old/new strides or by takind modulo.
+ */
+struct DropDims {
+  static constexpr int kMaxDims = 4;
+
+  int64_t div[kMaxDims];
+  int64_t mul[kMaxDims];
+  int64_t mod[kMaxDims];
+  int start = 2 * kMaxDims;
+
+  DropDims() = default;
+
+  /**
+   * Collapses adjacent groups of reduced/non-reduced dimensions.
+   * The input can have up to 64 dimensions, however, the result must havfe
+   * at most 8 groups of reduced/non-reduced dimensions.
+   *
+   * Example:
+   * ```
+   * in_shape:  [ 2, 3, 4, 5, 6 ]
+   * reduced:     ^        ^  ^
+   * axis_mask = 0b11001  (looks reversed - LSB is dim 0)
```

Comment:
What is the difference between `0b11001` and `0b00110`?

Reply:
`1` means the dimension is to be reduced, `0` the dimension is kept

Reply:
So I'm missing something, if it is what you say, then `in_shape:  [ 2, 3, 4, 5, 6 ]`, `axis_mask = 0b11001` should produce ` [ 2, 12, 5, 6 ]` while the example states `[ 2, 12, 30 ]`.

Reply:
Adjacent groups dimensions are collapsed if all of them are reduced or not reduced.

Reply:
So `0b11001` and `0b00110` should do the same, rigth?

Reply:
They should produce the same out_shape, but complementary mask and accordingly different div/mul/mod contents.
Say, you have 640x480x3 image. Whether you reduce HW and keep C or vice versa, you get 307200x3 tensor - however, what you do with it next is totally different.